### PR TITLE
gh-81322: support multiple separators in Stream.readuntil

### PR DESCRIPTION
**Pull request by bmerry**
Allow Stream.readuntil to take an iterable of separators and match any
of them. The earliest match endpoint wins (which ensures that results
are dependent on the chunking) and on ties shortest separator wins
(which only matters if the user has supplied a redundant set like
[b'\r\n', b'\n'] and the limit is reached).

It's also implemented for the deprecated StreamReader, just because the
code for the two implementations was the same except for one line and it
seemed easier to keep them in sync than leaving two different versions
to maintain.

https://bugs.python.org/issue37141

* Issue: gh-81322

### CODE REVIEW

**Comment by bmerry** [comment]
A few questions about this:
- I've implemented it for both StreamReader and Stream, because it seemed like having two identical (apart from `_ensure_can_read`) implementations is better for maintainability than two implementations that have drifted out of sync. But if one would prefer not to make any changes to StreamReader I can revert that. I suggest doing review discussion on the Stream implementation.
- Because the existing functionality is implemented using the general case there might be a very small loss of performance. Are there any benchmarks I should run to test that?
- For matches that end in the same position I arbitrarily decided that shortest should win for the purpose of LimitOverrunError, but it was an arbitrary choice. It could just as easily be longest-wins or first-in-list wins.

File: Lib/asyncio/streams.py
Comment (bmerry) [inline Lib/asyncio/streams.py]: Hmm, I see this can make offset negative if max_seplen > min_seplen. I'll add a test to catch it and add a fix.

File: Lib/asyncio/streams.py
Comment (brandtbucher) [inline Lib/asyncio/streams.py]: No need to wrap this in a lambda!

```suggestion
            separator = sorted(separator, key=len)
```

File: Lib/asyncio/streams.py
Comment (brandtbucher) [inline Lib/asyncio/streams.py]: Since you support any iterable, maybe leave "list" out of it?

```suggestion
            raise ValueError('Separator should contain at least one element')
```

File: Lib/asyncio/streams.py
Comment (brandtbucher) [inline Lib/asyncio/streams.py]: I'd leave the backticks here. Even though it's a list now, it still makes it clear that we're referring to the variable.

```suggestion
        # where there is no occurrence of any `separator`.
```

File: Lib/asyncio/streams.py
Comment (brandtbucher) [inline Lib/asyncio/streams.py]: Here too:

```suggestion
        # Loop until we find a `separator` in the buffer, exceed the buffer size,
```

File: Lib/asyncio/streams.py
Comment (brandtbucher) [inline Lib/asyncio/streams.py]: See above.

```suggestion
            raise ValueError('Separator should contain at least one element')
```

File: Lib/asyncio/streams.py
Comment (brandtbucher) [inline Lib/asyncio/streams.py]: See above.

```suggestion
        # where there is no occurrence of any `separator`.
```

File: Lib/asyncio/streams.py
Comment (brandtbucher) [inline Lib/asyncio/streams.py]: See above.

```suggestion
        # Loop until we find a `separator` in the buffer, exceed the buffer size,
```

**Review by brandtbucher — CHANGES_REQUESTED** [review]
Thanks for taking the time to draft this PR @bmerry, and welcome to CPython! :sunglasses:

I'm not *super* familiar with this module, but I found a couple of things that could be tidied up. Other than that, the code looks good! This will definitely need updated docs, though.

File: Lib/asyncio/streams.py
Comment (brandtbucher) [inline Lib/asyncio/streams.py]: Sorry, missed one. Here too!

```suggestion
            separator = sorted(separator, key=len)
```

**Review by 1st1 — CHANGES_REQUESTED** [review]
Unfortunately it was decided to revert the current streams implementation from 3.8.  See https://bugs.python.org/issue38242 for more details.  I'm really sorry, but we'll need to rebase this work on the new API we later add to 3.9 :(

**Comment by bmerry** [comment]
@1st1 not a problem - definitely worth getting things right before they go into stdlib. Do you think the revert of the new stream API will conflict much with my changes to the old StreamReader class? If not, perhaps the way forward is for me to remove my changes to `Stream`, address the PR comments on `StreamReader` and then hopefully it shouldn't need much work to rebase after `Stream` is removed?

**Comment by 1st1** [comment]
@bmerry I'm not yet sure. I'll be working on a revert later today or tomorrow and will probably know more.  I'd keep everything as is as it's not going to be an easy revert anyways.

**Comment by bmerry** [comment]
@1st1 it looks like you've done the Streams reversion now, and I've merged that into my branch. So hopefully this is ready for review now.

@brandtbucher thanks for the suggestions. I've applied all the ones that Github was still showing after the merge from master - let me know if anything got lost along the way.

File: Misc/NEWS.d/next/Library/2019-09-26-17-52-52.bpo-37141.onYY2-.rst
Comment (brandtbucher) [inline Misc/NEWS.d/next/Library/2019-09-26-17-52-52.bpo-37141.onYY2-.rst]:
```suggestion
Accept an iterable of separators in :meth:`asyncio.Stream.readuntil`, stopping
```

**Review by brandtbucher — CHANGES_REQUESTED** [review]
One more tiny clarification from me, otherwise this looks good. Still needs updated docs though!

File: Misc/NEWS.d/next/Library/2019-09-26-17-52-52.bpo-37141.onYY2-.rst
Comment (bmerry) [reply Misc/NEWS.d/next/Library/2019-09-26-17-52-52.bpo-37141.onYY2-.rst]: Done

**Comment by bmerry** [comment]
> Still needs updated docs though!

@brandtbucher which docs need updating?

**Review by brandtbucher — CHANGES_REQUESTED** [review]
One tiny change to the NEWS entry due to the API revert:

File: Misc/NEWS.d/next/Library/2019-09-26-17-52-52.bpo-37141.onYY2-.rst
Comment (brandtbucher) [inline Misc/NEWS.d/next/Library/2019-09-26-17-52-52.bpo-37141.onYY2-.rst]:
```suggestion
Accept an iterable of separators in :meth:`asyncio.StreamReader.readuntil`, stopping
```

**Comment by bmerry** [comment]
> The docs I'm referring to are located in Doc/library/asyncio-stream.rst

Thanks. I'd assumed the library docs were generated from the docstring, which is why I missed this. I'll update it.

**Review by brandtbucher — APPROVED** [review]

**Comment by brandtbucher** [comment]
Looks good to me! @1st1?

**Comment by bmerry** [comment]
@1st1 any news on this now that Python 3.9 has reached alpha?

**Comment by bmerry** [comment]
@1st1 any news on what streams will look like in Python 3.9?

**Comment by bmerry** [comment]
@1st1 @asvetlov any news on what streams will look like in Python 3.9?

**Comment by bmerry** [comment]
It looks like Python 3.9 isn't changing the stream API, so I've rebased the PR (although I assume it will only target 3.10 since 3.9 has reached beta). The rebasing turned into a bit of a mess so I ended up squashing the whole thing to make a single new commit.

**Comment by bmerry** [comment]
@brandtbucher any advice on how I can move this forward? There has been no response on it all year.

**Comment by brandtbucher** [comment]
Thanks for your patience @bmerry! I think @1st1 has probably just been busy.

While your work here still looks fine to me in terms of style and documentation, I personally know next to nothing about async stuff. Perhaps @aeros has some time to look at this (and perhaps merge it)?

**Comment by aeros** [comment]
Sorry @bmerry for the delay. There's still a strong long-term intention for an overhaul to the streams API, but due to other work being prioritized (such as exception groups, and by extension, task groups), we decided at the core sprint to delay work on a new streams API for now since it will be less impactful for most users. 

My thoughts are that supporting multiple separators in `readuntil()` is useful enough to consider as an enhancement to the current version of streams, particularly since it will remain functional for several versions before *eventually* being deprecated and removed. 

However, I'd like to get input from @1st1 and @asvetlov to decide if it's worthwhile from their perspective, and in general what our current plans are surrounding enhancements to the existing streams API (in light of the recent decision at the 2020 core sprint to delay work on it for now).

**Comment by gvanrossum** [comment]
Thanks, I will review in the coming week!

File: Misc/NEWS.d/next/Library/2019-09-26-17-52-52.bpo-37141.onYY2-.rst
Comment (bmerry) [inline Misc/NEWS.d/next/Library/2019-09-26-17-52-52.bpo-37141.onYY2-.rst:1]: Does the filename need to be changed to reflect the Github issue number instead of the BPO number?

Comment (gvanrossum) [reply Misc/NEWS.d/next/Library/2019-09-26-17-52-52.bpo-37141.onYY2-.rst:1]: I don't think so -- this is just used to make the files unique, at the next alpha release it will all be merged into a single 3.13.0a6.rst file.

File: Lib/asyncio/streams.py
Comment (gvanrossum) [inline Lib/asyncio/streams.py]:
```suggestion
        #   can safely consume max_seplen - 1 bytes.
```

**Review by gvanrossum — APPROVED** [review]
One comment nit, I'll merge that myself and then land it. Thanks for your contribution, and on behalf of the asyncio maintainers I apologize for the delay!

**Pull request merged**
